query/filter/key_builder.go

```diff
@@ -81,38 +158,66 @@ func (k *KeyBuilder) Build(filters []Filter, userDefinedKeys []*schema.Field) ([
 		queue = queue[1:]
 	}
 
+	// PrimaryKey is always a single plan with all the keys
+	if k.primaryKey {
+		combined := allKeys[0]
+		for _, plan := range allKeys[1:] {
+			combined.Keys = append(combined.Keys, plan.Keys...)
+		}
+		return []QueryPlan{combined}, nil
+	}
+
 	return allKeys, nil
 }
 
+func PKBuildIndexPartsFunc(name string, datatype schema.FieldType, value interface{}) []interface{} {
+	return []interface{}{value}
+}
+
 // KeyComposer needs to be implemented to have a custom Compose method with different constraints.
-type KeyComposer interface {
-	Compose(level []*Selector, userDefinedKeys []*schema.Field, parent LogicalOP) ([]keys.Key, error)
+type KeyComposer[F fieldable] interface {
+	Compose(level []*Selector, userDefinedKeys []F, parent LogicalOP) ([]QueryPlan, error)
 }
 
-// StrictEqKeyComposer is to generate internal keys only if the condition is equality on the fields that are part of
-// the schema and all these fields are present in the filters. The following rules are applied for StrictEqKeyComposer
-//   - The userDefinedKeys(indexes defined in the schema) passed in parameter should be present in the filter
-//   - For AND filters it is possible to build internal keys for composite indexes, for OR it is not possible.
+type (
+	KeyEncodingFunc     func(indexParts ...interface{}) (keys.Key, error)
+	BuildIndexPartsFunc func(fieldName string, datatype schema.FieldType, value interface{}) []interface{}
+)
+
+// StrictEqKeyComposer works in to ways to generate internal keys if the condition is equality.
+//  1. When `matchAll=true`, it will generate internal keys of equality on the fields of the schema if all these fields
+//     are present in the filters. The following rules are applied for StrictEqKeyComposer:
+//     - The userDefinedKeys(indexes defined in the schema) passed in parameter should be present in the filter
+//     - For AND filters it is possible to build internal keys for composite indexes, for OR it is not possible.
 //
-// So for OR filter an error is returned if it is used for indexes that are composite.
-type StrictEqKeyComposer struct {
+// 2. When `matchAll=false`, it will treat all userDefined as individual and generate an `$eq` query plan for each one that is found.
+//
+// For OR filter an error is returned if it is used for indexes that are composite.
+type StrictEqKeyComposer[F fieldable] struct {
```

Comment:
It might be better to split this out into two. `PKStrictEqKeyComposer` and `SecondaryEqComposer`

Reply:
Yes that makes sense.